File: inject-java/src/test/groovy/io/micronaut/aop/named/NamedFactory.java
/*
 * Copyright 2017-2020 original authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.micronaut.aop.named;

import io.micronaut.aop.Logged;
import io.micronaut.context.annotation.EachProperty;
import io.micronaut.context.annotation.Factory;
import io.micronaut.context.annotation.Parameter;
import io.micronaut.runtime.context.scope.Refreshable;

import javax.inject.Named;
import javax.inject.Singleton;

@Factory
public class NamedFactory {

    @EachProperty(value = "aop.test.named", primary = "default")
    @Refreshable
    NamedInterface namedInterface(@Parameter String name) {
        return () -> name;
    }


    @Named("first")
    @Logged
    @Singleton
    OtherInterface first() {
        return () -> "first";
    }

    @Named("second")
    @Logged
    @Singleton
    OtherInterface second() {
        return () -> "second";
    }

    @EachProperty("other.interfaces")
    OtherInterface third(Config config, @Parameter String name) {
        return () -> name;
    }
}

File: inject-kotlin/src/test/kotlin/io/micronaut/kotlin/processing/aop/named/NamedFactory.kt
package io.micronaut.kotlin.processing.aop.named

import io.micronaut.context.annotation.EachProperty
import io.micronaut.context.annotation.Factory
import io.micronaut.context.annotation.Parameter
import io.micronaut.kotlin.processing.aop.Logged
import io.micronaut.runtime.context.scope.Refreshable
import jakarta.inject.Named
import jakarta.inject.Singleton

@Factory
class NamedFactory {

    @EachProperty(value = "aop.test.named", primary = "default")
    @Refreshable
    fun namedInterface(@Parameter name: String): NamedInterface {
        return object : NamedInterface {
            override fun doStuff(): String {
                return name
            }
        }
    }

    @Named("first")
    @Logged
    @Singleton
    fun first(): OtherInterface {
        return object : OtherInterface {
            override fun doStuff(): String {
                return "first"
            }
        }
    }

    @Named("second")
    @Logged
    @Singleton
    fun second(): OtherInterface {
        return object : OtherInterface {
            override fun doStuff(): String {
                return "second"
            }
        }
    }

    @EachProperty("other.interfaces")
    fun third(config: Config, @Parameter name: String): OtherInterface {
        return object : OtherInterface {
            override fun doStuff(): String {
                return name
            }
        }
    }
}
